### PR TITLE
Docs on hook call order

### LINKED ISSUE #21069: Improve docs for Callbacks on access to all batch outputs
### 📚 Documentation

Hi Lightning team,

The docs encourage that the Callbacks should be "[NON-ESSENTIAL logic that is NOT required for your lightning module](https://lightning.ai/docs/pytorch/latest/extensions/callbacks.html#best-practices)".

However, in [the example of accessing all batch outputs](https://lightning.ai/docs/pytorch/latest/_modules/lightning/pytorch/callbacks/callback.html#Callback.on_train_epoch_end) in `on_train_epoch_end()` docs, the clearing of the memory is done inside the callback, making it *essential* to avoid memory build-up.

It would be nice to adapt this example so that memory clean-up does not depend on the Callback.

I would assume it is as easy as moving memory-clearing code to the LightningModule's `on_train_epoch_end()`, however, I couldn't find any documentation of what is being called first, the LightningModule's `on_train_epoch_end()` or the Callbacks... Maybe I have missed it, but if not, then it would be another improvement to the docs.

Cheers

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #21069

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: docs/source-pytorch/common/hooks.rst [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+with the source of each hook indicated:
+
+.. code-block:: text
+
+    Training Process Flow:
+
+    trainer.fit()
+    │
+    ├── setup(stage="fit")
+    │   └── [Callbacks only]
```
Comment (GdoongMathew) [inline docs/source-pytorch/common/hooks.rst:86]: just a suggestion. perhaps adding this could also be helpful? as it also shows one of the known issue https://github.com/Lightning-AI/pytorch-lightning/issues/19658

```
    ├── setup(stage="fit")
    │   └── [Callback.setup]
    │   └── [LightnintModule.configure_shared_model]
    │   └── [LightnintModule.configure_model]
    │   └── [Strategy.setup]
```

File: docs/source-pytorch/common/hooks.rst [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    │
+    ├── on_fit_start()
+    │   ├── [Callbacks]
+    │   ├── [LightningModule]
+    │   └── [Strategy]
+    │
+    ├── on_sanity_check_start()
```
Comment (GdoongMathew) [inline docs/source-pytorch/common/hooks.rst:93]: perhaps we could also add when the checkpoint and state_dict are load?
```
on_fit_start()

if strategy.restore_checkpoint_after_setup:
    [LightningModule.on_load_checkpoint]
    [Strategy.load_model_state_dict()

[optimizer.load_state_dict]
[lr_scheduler.load_state_dict]

on_sanity_check_start()
```

File: docs/source-pytorch/common/hooks.rst [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    │   ├── [LightningModule]
+    │   └── [Strategy]
+    │
+    ├── on_fit_end()
+    │   ├── [Callbacks]
+    │   ├── [LightningModule]
+    │   └── [Strategy]
+    │
+    └── teardown(stage="fit")
+        └── [Callbacks only]
```
Comment (GdoongMathew) [inline docs/source-pytorch/common/hooks.rst:221]: I believe that the lightning module also has a `teardown` callback here.
```
    └── teardown(stage="fit")
        └── [Callback]
        └── [LightningModule]
```

File: docs/source-pytorch/common/hooks.rst [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    │   ├── [Callbacks]
+    │   ├── [LightningModule]
+    │   └── [Strategy]
+    │
+    ├── on_fit_end()
+    │   ├── [Callbacks]
+    │   ├── [LightningModule]
+    │   └── [Strategy]
+    │
+    └── teardown(stage="fit")
```
Comment (GdoongMathew) [inline docs/source-pytorch/common/hooks.rst:220]: from my personal code inspect, I believe that the strategy.teardown should be called before `on_fit_end` stage, and also the strategy class does not have a `on_fit_end` callback.

```
strategy.teardown()
     |--LightningModule.cpu()
on_fit_end()
     |--Callback.on_fit_end()
     |--LightningModule.on_fit_end()
Callback.teardown()
LightningModule.teardown()
```

File: docs/source-pytorch/common/hooks.rst [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    │   │   │   ├── [Callbacks]
+    │   │   │   ├── [LightningModule]
+    │   │   │   └── [Strategy]
+    │   │   │
+    │   │   ├── on_before_zero_grad()
+    │   │   │   ├── [Callbacks]
+    │   │   │   └── [LightningModule]
+    │   │   │
+    │   │   ├── [Forward Pass - training_step()]
+    │   │   │   └── [Strategy only]
```
Comment (GdoongMathew) [inline docs/source-pytorch/common/hooks.rst:151]: I think the `training step` is called before the `on_before_zero_grad` is called?
```
strategy.training_step()
    | -- LightningModule.training_step()
on_before_zero_grad()
    |-- Callback.on_before_zero_grad()
    |-- LightningModule.on_before_zero_grad()
    |-- LightningModule.optimizer_zero_grad()
```

File: docs/source-pytorch/common/hooks.rst [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+    │   │   ├── [Backward Pass]
+    │   │   │   └── [Strategy only]
+    │   │   │
+    │   │   ├── on_after_backward()
+    │   │   │   ├── [Callbacks]
+    │   │   │   └── [LightningModule]
+    │   │   │
+    │   │   ├── on_before_optimizer_step()
+    │   │   │   ├── [Callbacks]
+    │   │   │   └── [LightningModule]
```
Comment (GdoongMathew) [inline docs/source-pytorch/common/hooks.rst:166]: perhaps also add `configure_gradient_clipping` here?
```
    │   │   ├── on_before_optimizer_step()
    │   │   │   ├── [Callbacks]
    │   │   │   ├── [LightningModule]
    │   │   │   └── [LightningModule.configure_gradient_clipping]
```

File: docs/source-pytorch/common/hooks.rst [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+    │   │   │   └── [Strategy]
+    │   │   └── on_validation_end()
+    │   │       ├── [Callbacks]
+    │   │       ├── [LightningModule]
+    │   │       └── [Strategy]
+    │   │
+    │   └── on_train_epoch_end() **SPECIAL CASE**
+    │       ├── [Callbacks - Non-monitoring only]
+    │       ├── [LightningModule]
+    │       └── [Callbacks - Monitoring only]
```
Comment (GdoongMathew) [inline docs/source-pytorch/common/hooks.rst:206]: Perhaps `Non-monitoring` could be changed to `Checkpoint` & `EarlyStopping` to be more specific.

**Review by Borda — APPROVED** [review]

File: docs/source-pytorch/common/hooks.rst [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+- **Strategy**: Custom strategies for distributed training.
+
+Importantly, because logic can be place in the same hook but in different places the call order of hooks is in
+important to understand. The following order is always used:
+
+1. Callbacks, called in the order they are passed to the Trainer.
+2. ``LightningModule``
+3. Strategy
+
+.. testcase::
```
Comment (Borda) [inline docs/source-pytorch/common/hooks.rst]:
```suggestion
.. testcode::
```

**Comment by Borda** [comment]
@GdoongMathew, all good call. How about landing this as a baseline, and you can send just another PR with your suggestions for improvement? :)

**Comment by GdoongMathew** [comment]
From my previous experience, since Lightning uses the closure function in the optimizer, it can make tracing the optimizer step a bit difficult. Maybe we could also highlight that here?
```
            # closure function for optimizer
            strategy.training_step()
            └── LightningModule.training_step()

            callback.on_before_zero_grad()
            LightningModule.on_before_zero_grad()
            LightningModule.optimizer_zero_grad()
            strategy.backward()
            ├── callback.on_before_backward()
            ├── LightningModule.on_before_backward()
            ├── LightningModule.backward()
            │   └── loss.backward()
            ├── callback.on_after_backward()
            └── LightningModule.on_after_backward()

            callback.on_before_optimizer_step()
            LightningModule.on_before_optimizer_step()
            LightningModule.configure_gradient_clipping()
    
            optimizer.step(closure)
```

but this also makes my personal note a bit messier as well....

**Comment by GdoongMathew** [comment]
@Borda sure!! will open another PR for those suggestions. thanks~

**Comment by Borda** [comment]
> will open another PR for those suggestions. thanks~

@GdoongMathew would you open for a quick chat? Can you share some connection, or ping me https://www.linkedin.com/in/jirka-borovec/

**Comment by Borda** [comment]
@GdoongMathew, it has landed, so now you can send your suggestions with improvments 🚀

**Review by GdoongMathew — APPROVED** [review]
